File: Tarea9 (HeapSort)/Tarea9.cpp
#include <iostream>
#include <vector>
#include <algorithm>

void heapify(std::vector<int>& arr, int n, int i) {
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;

    if (left < n && arr[left] > arr[largest]) {
        largest = left;
    }

    if (right < n && arr[right] > arr[largest]) {
        largest = right;
    }

    if (largest != i) {
        std::swap(arr[i], arr[largest]);
        heapify(arr, n, largest);
    }
}

void heapsort(std::vector<int>& arr) {
    int n = arr.size();

    for (int i = n / 2 - 1; i >= 0; i--) {
        heapify(arr, n, i);
    }

    for (int i = n - 1; i > 0; i--) {
        std::swap(arr[0], arr[i]);
        heapify(arr, i, 0);
    }
}

void printArray(const std::vector<int>& arr) {
    for (int val : arr) {
        std::cout << val << " ";
    }
    std::cout << std::endl;
}

int main() {
    std::vector<int> lista = {12, 11, 13, 5, 6, 7};
    std::cout << "Antes de ordenar los elementos del array son: ";
    printArray(lista);
    heapsort(lista);
    std::cout << "\nDespués de ordenar el arreglo: ";
    printArray(lista);
    return 0;
}

File: Tarea9 (HeapSort)/Tarea9.cs
using System;

public class Tarea9
{
    public static void Heapify(int[] arr, int n, int i)
    {
        int largest = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        if (left < n && arr[left] > arr[largest])
        {
            largest = left;
        }

        if (right < n && arr[right] > arr[largest])
        {
            largest = right;
        }

        if (largest != i)
        {
            (arr[i], arr[largest]) = (arr[largest], arr[i]);
            Heapify(arr, n, largest);
        }
    }

    public static void Sort(int[] arr)
    {
        int n = arr.Length;

        for (int i = n / 2 - 1; i >= 0; i--)
        {
            Heapify(arr, n, i);
        }

        for (int i = n - 1; i > 0; i--)
        {
            (arr[0], arr[i]) = (arr[i], arr[0]);
            Heapify(arr, i, 0);
        }
    }

    public static void Main(string[] args)
    {
        int[] lista = { 12, 11, 13, 5, 6, 7 };
        Console.WriteLine("Antes de ordenar los elementos del array son: [" + string.Join(", ", lista) + "]");
        Sort(lista);
        Console.WriteLine("\nDespués de ordenar el arreglo: [" + string.Join(", ", lista) + "]");
    }
}
